Mathlib/LinearAlgebra/RootSystem/OfBilinear.lean

```diff
@@ -0,0 +1,193 @@
+These conditions are what we need when describing reflection as a map taking `y` to
+`y - 2 • (B x y) / (B x x) • x`. -/
+structure IsReflective (B : M →ₗ[R] M →ₗ[R] R) (x : M) : Prop :=
+  regular : IsRegular (B x x)
+  dvd_two_mul : ∀ y, B x x ∣ 2 * B x y
+
+variable (B : M →ₗ[R] M →ₗ[R] R) {x : M}
+
+namespace IsReflective
+
+lemma of_dvd_two [IsDomain R] [NeZero (2 : R)] (hx : B x x ∣ 2) :
+    IsReflective B x where
+  regular := isRegular_of_ne_zero <| fun contra ↦ by simp [contra, two_ne_zero (α := R)] at hx
+  dvd_two_mul y := hx.mul_right (B x y)
+
+variable (hx : IsReflective B x)
+
+/-- The coroot attached to a reflective vector. -/
+def coroot : M →ₗ[R] R where
+  toFun y := (hx.2 y).choose
+  map_add' a b := by
+    refine hx.1.1 ?_
+    simp only
+    rw [← (hx.2 (a + b)).choose_spec, mul_add, ← (hx.2 a).choose_spec, ← (hx.2 b).choose_spec,
+      map_add, mul_add]
+  map_smul' r a := by
+    refine hx.1.1 ?_
+    simp only [RingHom.id_apply]
+    rw [← (hx.2 (r • a)).choose_spec, smul_eq_mul, mul_left_comm, ← (hx.2 a).choose_spec, map_smul,
+      two_mul, smul_eq_mul, two_mul, mul_add]
+
+@[simp]
+lemma apply_self_mul_coroot_apply {y : M} : B x x * coroot B hx y = 2 * B x y :=
+  (hx.dvd_two_mul y).choose_spec.symm
+
+@[simp]
+lemma smul_coroot : B x x • coroot B hx = 2 • B x := by
+  ext y
+  simp [smul_apply, smul_eq_mul, nsmul_eq_mul, Nat.cast_ofNat, apply_self_mul_coroot_apply]
+
+@[simp]
+lemma coroot_apply_self : coroot B hx x = 2 :=
+  hx.regular.left <| by simp [mul_comm _ (B x x)]
+
+lemma isOrthogonal_reflection (hSB : LinearMap.IsSymm B) :
+    B.IsOrthogonal (Module.reflection (coroot_apply_self B hx)) := by
+  intro y z
+  simp only [LinearEquiv.coe_coe, reflection_apply, LinearMap.map_sub, map_smul, sub_apply,
+    smul_apply, smul_eq_mul]
+  refine hx.1.1 ?_
+  simp only [mul_sub, ← mul_assoc, apply_self_mul_coroot_apply]
+  rw [sub_eq_iff_eq_add, ← hSB x y, RingHom.id_apply, mul_assoc _ _ (B x x), mul_comm _ (B x x),
+    apply_self_mul_coroot_apply]
+  ring
+
+lemma reflective_reflection_reflective (B : M →ₗ[R] M →ₗ[R] R) (hSB : LinearMap.IsSymm B) {x y : M}
+    (hx : IsReflective B x) (hy : IsReflective B y) :
+    IsReflective B (Module.reflection (coroot_apply_self B hx) y) := by
+  constructor
+  · rw [← LinearEquiv.coe_coe, isOrthogonal_reflection B hx hSB]
+    exact hy.1
+  · intro z
+    have hz : Module.reflection (coroot_apply_self B hx)
+        (Module.reflection (coroot_apply_self B hx) z) = z := by
+      exact (LinearEquiv.eq_symm_apply (Module.reflection (coroot_apply_self B hx))).mp rfl
+    rw [← hz, ← LinearEquiv.coe_coe, isOrthogonal_reflection B hx hSB,
+      isOrthogonal_reflection B hx hSB]
+    exact hy.2 _
+
+@[simp]
+lemma dual_apply [IsReflexive R M] (B : M →ₗ[R] M →ₗ[R] R) {x y : M} (hx : IsReflective B x) :
+    IsReflexive.toPerfectPairingDual (R := R) (coroot B hx) y =
+    (coroot B hx) y :=
+  rfl
+
```

Comment:
I think we should drop this:
```suggestion
```
and instead add the following just below the definition of `IsReflexive.toPerfectPairingDual` in `Mathlib.LinearAlgebra.PerfectPairing`:
```lean
@[simp]
lemma IsReflexive.toPerfectPairingDual_apply {f : Dual R M} {x : M} :
    IsReflexive.toPerfectPairingDual (R := R) f x = f x :=
  rfl
```